compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -1279,6 +1288,136 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
+                  //    - If both source and destination arrays are primitive VT
+                  //        - (1) Both arrays are not flattened, there is no need to do anything
+                  //        - (2) Both arrays are flattened, elementSize needs to be updated in order to
+                  //          use arraycopy instructions
+                  //        - (3) Either of the arrays might or might be flattened, System.arraycopy
+                  //          should not be transformed into arraycopy instructions.
+                  //
+                  if (isValueTypeArrayFlatteningEnabled)
+                     {
+                     if (trace())
+                        traceMsg(comp(), "%s: n%dn %p isArrayElementFlattened dst %d src %d\n", __FUNCTION__, node->getGlobalIndex(), node,
+                           isArrayElementFlattened(dstObject), isArrayElementFlattened(srcObject));
+
+                     if ((isArrayElementFlattened(dstObject) == TR_yes) &&
+                         (isArrayElementFlattened(srcObject) == TR_yes))
+                        {
+                        areBothArraysFlattenedPrimitiveValueType = true;
+                        elementSize = TR::Compiler->cls.flattenedArrayElementSize(comp(), dstObject->getClass());
+
+                        if (trace())
+                           traceMsg(comp(), "%s: n%dn %p elementSize %d\n", __FUNCTION__, node->getGlobalIndex(), node, elementSize);
+                        }
+                     else if ((isArrayElementFlattened(dstObject) != TR_no) ||
+                              (isArrayElementFlattened(srcObject) != TR_no))
+                        {
+                        transformTheCall = false;
+                        }
+                     }
+                  }
+               else // isSrcArrayCompTypePrimitiveValueType == TR_no or TR_maybe
+                  {
+                  // The destination is primitive VT array and the source might or might not
+                  // be primitive VT array, do not transform because we need to do null store check and
+                  // consider if the arrays are flattened
+                  transformTheCall = false;
+                  }
+               break;
+               }
+            case TR_maybe:
+               {
+               if (isSrcArrayCompTypePrimitiveValueType != TR_no)
+                  {
+                  // The source might or might not be a primitive VT array and the target might or might not
+                  // be a primitive VT array, add runtime check before the arraycopy instructions
+                  doRuntimeNullRestrictedTest = true;
+                  }
+               // else: The source is not primitive VT array and the target might or might not a primitive VT,
+               //       nothing needs to be done here because:
+               //       - If in runtime the destination is primitive VT array, it should fail array store check
```

Comment:
I'll run experiment to double check this case

Reply:
> What if the source array is a fixed array of type java/lang/Object and the target turns out to be a flattened primitive value type array at runtime? 

I reproduced issue with this case. Good catch! I think if destination array might or might not be the primitive vt array, we need to add runtime check regardless of the source array type. 

```c++
            case TR_maybe:
               {
               if (isSrcArrayCompTypePrimitiveValueType != TR_no)
                  {
                  // isSrcArrayCompTypePrimitiveValueType either equals to TR_yes or TR_maybe 
                  // The source might or might not be a primitive VT array and the target might or might not
                  // be a primitive VT array, add runtime check before the arraycopy instructions
                  doRuntimeNullRestrictedTest = true;
                  }
               // else: The source is not primitive VT array and the target might or might not a primitive VT,
               //       nothing needs to be done here because:
               //       - If in runtime the destination is primitive VT array, it should fail array store check
               //       - If in runtime the destination is not primitive VT array, it should function as it is today
               break;
               }
```
==>
```c++
            case TR_maybe:
               {
               doRuntimeNullRestrictedTest = true;
               break;
               }
```

Reply:
Adressed in 71ddb70